curriculum/challenges/english/07-scientific-computing-with-python/learn-string-manipulation-by-building-a-cipher/65521fc818947e800bffe48a.md

```diff
@@ -1,44 +1,47 @@
 ---
 id: 65521fc818947e800bffe48a
-title: Step 23
+title: Step 27
 challengeType: 20
-dashedName: step-23
+dashedName: step-27
 ---
 
 # --description--
 
-Before printing the current character, declare a variable called `index` and assign the value returned by `alphabet.find(char)` to this variable.
+Inside the `for` loop, before printing the current character, declare a variable called `index` and assign the value returned by `alphabet.find(char)` to this variable.
 
 # --hints--
 
-You should declare a new variable named `index`.
+You should declare a new variable named `index` at the beginning of your `for` loop.
 
 ```js
-({ test: () => assert(__userGlobals.has("index")) })
+const commentless_code = __helpers.python.removeComments(code);
+const {block_body} = __helpers.python.getBlock(commentless_code, /for\s+char\s+in\s+text\s*/);
+assert(block_body.match(/^\s*index\s*=/));
 ```
 
-You should assign `alphabet.find(char)` to your new variable.
+You should assign `alphabet.find(char)` to your new `index` variable.
 
 ```js
-({ test: () => assert.match(code, /index\s*=\s*alphabet\.find\s*\(\s*char\s*\)/) })
+const commentless_code = __helpers.python.removeComments(code);
+const {block_body} = __helpers.python.getBlock(commentless_code, /for\s+char\s+in\s+text\s*/);
+assert(block_body.match(/index\s*=\s*alphabet\.find\s*\(\s*char\s*\)\s*(#.*)?$/m));
 ```
 
-You should declare your new `index` variable at the beginning of your `for` loop.
+Your code contains invalid syntax and/or invalid indentation.
```

Comment:
For the wrong indentation, I got the error `Your code raised an error before any tests could run. Please fix it and try again.` instead of `Your code contains invalid syntax and/or invalid indentation.` Is this the expected behaviour?

Reply:
The sentence `Your code contains invalid syntax and/or invalid indentation.` never shows up. If the code is correct the test passes, otherwise it fails and you see `Your code raised an error before any tests could run. Please fix it and try again`. The only purpose of that test is to run the code and check that everything is fine.
Many hints does not show without doing it, and considering the forum posts I think it's vital to show every hint at least at the beginning of the project.